Dev/deploy from catalog

janstaelensskyline opened the pull request.

janstaelensskyline committed:
Add project files.

janstaelensskyline committed:
Renamed Dataminer to DataMiner

janstaelensskyline committed:
Finished FromCatalog command and code.
TODO: readme

janstaelensskyline committed:
Updated README

janstaelensskyline committed:
Rename Artifacts folder to  DataMinerArtifacts

MichielOda commented on CICD.Tools.DataMinerDeploy.LibTests/CICD.Tools.DataMinerDeploy.LibTests.csproj:
```diff
@@ -0,0 +1,27 @@
+    <PackageReference Include="coverlet.collector" Version="3.1.2" />
+    <PackageReference Include="Serilog" Version="3.1.1" />
+    <PackageReference Include="Serilog.Sinks.Console" Version="5.0.1" />
+  </ItemGroup>
```
```suggestion
  <ItemGroup>
    <PackageReference Include="Extensions.Logging.Mock" Version="1.2.1" />
    <PackageReference Include="FluentAssertions" Version="6.12.0" />
    <PackageReference Include="Microsoft.NET.Test.Sdk" Version="17.8.0" />
    <PackageReference Include="Moq" Version="4.20.70" />
    <PackageReference Include="MSTest.TestAdapter" Version="3.1.1" />
    <PackageReference Include="MSTest.TestFramework" Version="3.1.1" />
    <PackageReference Include="coverlet.collector" Version="6.0.0" />
    <PackageReference Include="Serilog.Sinks.Console" Version="5.0.1" />
  </ItemGroup>
```

MichielOda commented on CICD.Tools.DataminerDeploy/Program.cs:
```diff
@@ -0,0 +1,167 @@
+				IsRequired = true
+			};
+
+			var FromArtifact = new Command("FromArtifact", "Deploys a specific package from a local .dmapp to a DataMiner agent.")
```
Please use camelCase for variables. I would suggest to change the command to 'from-artifact'.

janstaelensskyline replied:
camelcase OK, I would like to avoid special characters for the commandline though. This to avoid wierd behavior on groovy, linux, windows, concourse, github, azure, ... wherever they execute from. The less special tokens, the more robust.

MichielOda replied:
![image](https://github.com/SkylineCommunications/Skyline.DataMiner.CICD.Tools.DataMinerDeploy/assets/97458010/d1b57895-ffe0-406f-b69d-eafc7f587d97)
The main command already has dashes as well. Even dotnet commands contain dashes. 
![image](https://github.com/SkylineCommunications/Skyline.DataMiner.CICD.Tools.DataMinerDeploy/assets/97458010/e63b61bd-d5a1-4605-9dab-79e5d633989d)

janstaelensskyline replied:
I still prefer to use PascalCase for the subcommand. It looks more natural to consider that the same like calling a Method. The main command dataminer-package-deploy is better differentiated from its subcommand then.

janstaelensskyline replied:
As discussed in person. Microsoft has official guidelines: https://learn.microsoft.com/en-us/dotnet/standard/commandline/syntax#design-guidance

as such we'll follow those standards and I've indeed adjusted the arguments to use Kebab-Case names.

MichielOda commented on CICD.Tools.DataminerDeploy/Program.cs:
```diff
@@ -0,0 +1,167 @@
+			var logger = loggerFactory.CreateLogger("Skyline.DataMiner.CICD.Tools.DataMinerDeploy");
+
+			IArtifact artifact;
+			if (string.IsNullOrWhiteSpace(dmCatalogToken))
```
```suggestion
			if (String.IsNullOrWhiteSpace(dmCatalogToken))
```

MichielOda commented on CICD.Tools.DataminerDeploy/Program.cs:
```diff
@@ -0,0 +1,167 @@
+			}
+			else if (deployTimeout == 0)
+			{
+				deployTimeout = int.MaxValue; // MaxValue
```
```suggestion
				deployTimeout = Int32.MaxValue; // MaxValue
```

MichielOda commented on CICD.Tools.DataminerDeploy/Program.cs:
```diff
@@ -0,0 +1,167 @@
+
+			deployTimeout.SetDefaultValue(-1);
+
+			var FromCatalog = new Command("FromCatalog", "Deploys a specific package from the cloud to a cloud-connected DataMiner agent. Currently only supports private artifacts uploaded using a key from the organization.")
```
Please use camelCase for variables. I would also suggest to change the sub command to 'from-catalog'.

janstaelensskyline replied:
camelcase OK, I would like to avoid special characters for the commandline though. This to avoid wierd behavior on groovy, linux, windows, concourse, github, azure, ... wherever they execute from. The less special tokens, the more robust.

MichielOda commented on CICD.Tools.DataminerDeploy/LICENSE.txt:
Update to new license with the metrics clause

MichielOda commented on CICD.Tools.DataminerDeploy/CICD.Tools.DataminerDeploy.csproj:
```diff
@@ -0,0 +1,41 @@
+		<AssemblyName>Skyline.DataMiner.CICD.Tools.DataMinerDeploy</AssemblyName>
+		<PackageVersion>1.0.1</PackageVersion>
+		<Version>1.0.1</Version>
+		<PackageTags>Skyline;DataMiner</PackageTags>
```
```suggestion
		<Nullable>disable</Nullable>
		<AssemblyName>Skyline.DataMiner.CICD.Tools.DataMinerDeploy</AssemblyName>
		<RootNamespace>Skyline.DataMiner.CICD.Tools.DataMinerDeploy</RootNamespace>
		<PackageTags>Skyline;DataMiner;CICD</PackageTags>
```

MichielOda commented on CICD.Tools.DataminerDeploy/CICD.Tools.DataminerDeploy.csproj:
```diff
@@ -0,0 +1,41 @@
+		<PackageReference Include="Serilog.Sinks.Console" Version="5.0.1" />
+		<PackageReference Include="Skyline.DataMiner.CICD.FileSystem" Version="1.0.0.2" />
+		<PackageReference Include="System.CommandLine" Version="2.0.0-beta4.22272.1" />
+	</ItemGroup>
```
```suggestion
	<ItemGroup>
		<PackageReference Include="Serilog.Extensions.Logging" Version="8.0.0" />
		<PackageReference Include="Serilog.Sinks.Console" Version="5.0.1" />
		<PackageReference Include="Skyline.DataMiner.CICD.FileSystem" Version="1.0.1" />
		<PackageReference Include="System.CommandLine" Version="2.0.0-beta4.22272.1" />
	</ItemGroup>
```

MichielOda commented on Skyline.DataMiner.CICD.Tools.DataminerDeploy.Lib/CICD.Tools.DataminerDeploy.Lib.csproj:
```diff
@@ -0,0 +1,36 @@
+		<AssemblyName>Skyline.DataMiner.CICD.Tools.DataMinerDeploy.Lib</AssemblyName>
+		<PackageVersion>1.0.1</PackageVersion>
+		<Version>1.0.1</Version>
+		<PackageTags>Skyline;DataMiner</PackageTags>
```
```suggestion
		<Nullable>disable</Nullable>
		<AssemblyName>Skyline.DataMiner.CICD.Tools.DataMinerDeploy.Lib</AssemblyName>
		<RootNamespace>Skyline.DataMiner.CICD.Tools.DataMinerDeploy.Lib</RootNamespace>
		<PackageTags>Skyline;DataMiner;CICD</PackageTags>
```

MichielOda commented on Skyline.DataMiner.CICD.Tools.DataminerDeploy.Lib/CICD.Tools.DataminerDeploy.Lib.csproj:
```diff
@@ -0,0 +1,36 @@
+
+	<ItemGroup>
+	  <PackageReference Include="Microsoft.Extensions.Logging" Version="8.0.0" />
+	  <PackageReference Include="Microsoft.Rest.ClientRuntime" Version="2.3.24" />
```
Package is deprecated. Apparently we need to use Azure.Core instead.

janstaelensskyline replied:
It doesn't recognize HttpOperationResponse  which is part of the generated code coming from swagger when we use Azure.Core. 

Would prefer to stick to the clientRunTime for the first release at least. That code is well-tested from usage in the GitHub deploy action.

https://stackoverflow.com/questions/77401708/how-to-migrate-from-microsoft-rest-clientruntime-to-azure-core

MichielOda commented on CICD.Tools.DataMinerDeploy.LibTests/DeploymentFactoryTests.cs:
Unit tests fail for me with this error: Requested registry access is not allowed.

janstaelensskyline replied:
You need to run your visual studio as administrator

MichielOda commented on Skyline.DataMiner.CICD.Tools.DataminerDeploy.Lib/CatalogService/CatalogServiceFactory.cs:
```diff
@@ -0,0 +1,39 @@
+				apiBaseUrl = "https://api.dataminer.services/api";
+			}
+
+			httpClient.BaseAddress = new Uri($"{apiBaseUrl}");
```
```suggestion
			httpClient.BaseAddress = new Uri(apiBaseUrl);
```

MichielOda commented on Skyline.DataMiner.CICD.Tools.DataminerDeploy.Lib/CatalogService/CatalogServiceFactory.cs:
```diff
@@ -0,0 +1,39 @@
+		/// <returns>An instance of <see cref="ICatalogService"/> to communicate with the Skyline DataMiner Catalog (https://catalog.dataminer.services/).</returns>
+		public static ICatalogService CreateWithHttp(HttpClient httpClient, ILogger logger)
+		{
+			var environment = Environment.GetEnvironmentVariable("Skyline-deploy-action-namespace");
```
Generally environment variables are upper case: [wikipedia](https://en.wikipedia.org/wiki/Environment_variable#:~:text=It%20is%20conventional,2%2C%20and%20Windows.)

janstaelensskyline replied:
done

MichielOda commented on Skyline.DataMiner.CICD.Tools.DataminerDeploy.Lib/CatalogService/HttpCatalogService.cs:
```diff
@@ -0,0 +1,135 @@
+			try
+			{
+				_logger.LogDebug($"Deploying {artifactIdentifier}");
+				res = await _deployArtifactApi.DeployArtifactWithApiKeyFunctionWithHttpMessagesAsync(new DeployArtifactAsSystemForm(artifactIdentifier), key);
```
Async method has an option to pass the cancellation token

janstaelensskyline replied:
OK

MichielOda commented on Skyline.DataMiner.CICD.Tools.DataminerDeploy.Lib/CatalogService/HttpCatalogService.cs:
```diff
@@ -0,0 +1,135 @@
+					_logger.LogDebug($"Deployment {deploymentId} started...");
+					return new DeployingPackage(artifactIdentifier, deploymentId);
+				}
+				throw new InvalidOperationException("Received an invalid deployment ID");
```
Style code: blank line after bracket (other locations as well)

janstaelensskyline replied:
OK

MichielOda commented on Skyline.DataMiner.CICD.Tools.DataminerDeploy.Lib/CatalogService/HttpCatalogService.cs:
```diff
@@ -0,0 +1,135 @@
+				throw new InvalidOperationException($"The GetDeployedPackage API returned a response with status code {res.Response.StatusCode}");
+			}
+
+			var responseContent = string.Empty;
```
```suggestion
			var responseContent = String.Empty;
```

MichielOda commented on Skyline.DataMiner.CICD.Tools.DataminerDeploy.Lib/DataMinerArtifacts/CatalogArtifact.cs:
Same feedback for env variables as before. In addition, maybe make it a const somewhere so that if it needs to be changed, you don't need to update each location separately.

MichielOda commented on Skyline.DataMiner.CICD.Tools.DataminerDeploy.Lib/DataMinerArtifacts/CatalogArtifact.cs:
```diff
@@ -0,0 +1,168 @@
+					var encryptedKey = WinEncryptedKeys.Lib.Keys.RetrieveKey("dmcatalogtoken_encrypted");
+					if (encryptedKey != null)
+					{
+						string keyFromWinEncryptedKeys = new System.Net.NetworkCredential(string.Empty, encryptedKey).Password;
```
```suggestion
						string keyFromWinEncryptedKeys = new System.Net.NetworkCredential(String.Empty, encryptedKey).Password;
```

MichielOda commented on Skyline.DataMiner.CICD.Tools.DataminerDeploy.Lib/DataMinerArtifacts/CatalogArtifact.cs:
```diff
@@ -0,0 +1,168 @@
+						if (!String.IsNullOrWhiteSpace(keyFromWinEncryptedKeys))
+						{
+							_logger.LogDebug("OK: Found token in Env Variable: 'dmcatalogtoken_encrypted' created by WinEncryptedKeys.");
+							keyFromEnv = keyFromWinEncryptedKeys;
```
Missing return? As you got the key so no need to execute the other part below.

janstaelensskyline replied:
Not a missing return. It should prioritize the key from dmcatalogtoken if it exists. so overwrites it. I also want to debug log if that override happens so the user is aware. I've added a comment to make that more clear

MichielOda replied:
I don't really understand why it should be able to be overridden? Currently the safer option (encrypted) is less preferred then the unencrypted one?

janstaelensskyline replied:
As discussed in person. This is mostly from a user perspective, in most pipeline technology like jenkins/azure/github the use of regular 'unencrypted' environment variables is the industry standard. So we'll give priority to that above our custom tool that add additional encryption.
That way we can hopefully avoid confusion or unexpected behavior when a user manually runs the tool under a specific environment or by passing along the credentials.

So currently priority is given as follows:
1. your own provided credentials through arguments
2. your credentials provided through environment variable
3. your credentials provided through our custom Skyline.DataMiner.CICD.Tools.WinEncryptedKeys tool

MichielOda commented on Skyline.DataMiner.CICD.Tools.DataminerDeploy.Lib/DataMinerArtifacts/IArtifact.cs:
```diff
@@ -0,0 +1,23 @@
+﻿using System;
+using System.Threading.Tasks;
```
usings inside namespace

janstaelensskyline replied:
Done, do you happen to know if there's a way to change the default behavior of VS to put them under the namespace iso above it?

MichielOda replied:
https://stackoverflow.com/a/66772071

janstaelensskyline replied:
Thanks!

MichielOda commented on Skyline.DataMiner.CICD.Tools.DataminerDeploy.Lib/DeploymentFactory.cs:
```diff
@@ -0,0 +1,75 @@
+﻿using Microsoft.Extensions.Logging;
```
using inside namespace

janstaelensskyline replied:
done

MichielOda commented on Skyline.DataMiner.CICD.Tools.DataminerDeploy.Lib/Models/DeployedPackage.cs:
```diff
@@ -0,0 +1,12 @@
+			Status = status;
+		}
+
+		public string Status { get; set; }
```
```suggestion
		public string Status { get; }
```

janstaelensskyline committed:
Update CICD.Tools.DataMinerDeploy.LibTests/CICD.Tools.DataMinerDeploy.LibTests.csproj

Co-authored-by: Michiel Oda <97458010+MichielOda@users.noreply.github.com>

janstaelensskyline committed:
Update CICD.Tools.DataminerDeploy/Program.cs

Co-authored-by: Michiel Oda <97458010+MichielOda@users.noreply.github.com>

janstaelensskyline committed:
Update CICD.Tools.DataminerDeploy/Program.cs

Co-authored-by: Michiel Oda <97458010+MichielOda@users.noreply.github.com>

janstaelensskyline committed:
Update Skyline.DataMiner.CICD.Tools.DataminerDeploy.Lib/DataMinerArtifacts/CatalogArtifact.cs

Co-authored-by: Michiel Oda <97458010+MichielOda@users.noreply.github.com>

janstaelensskyline committed:
Update Skyline.DataMiner.CICD.Tools.DataminerDeploy.Lib/Models/DeployedPackage.cs

Co-authored-by: Michiel Oda <97458010+MichielOda@users.noreply.github.com>

janstaelensskyline committed:
Applied review remarks.
Environment variables adjusted to

DATAMINER_CATALOG_TOKEN_ENCRYPTED
DATAMINER_CATALOG_TOKEN

janstaelensskyline committed:
Merge branch 'dev/DeployFromCatalog' of https://github.com/SkylineCommunications/Skyline.DataMiner.CICD.Tools.DataMinerDeploy into dev/DeployFromCatalog

janstaelensskyline committed:
Update CICD.Tools.DataminerDeploy/CICD.Tools.DataminerDeploy.csproj

Co-authored-by: Michiel Oda <97458010+MichielOda@users.noreply.github.com>

janstaelensskyline committed:
Update CICD.Tools.DataminerDeploy/CICD.Tools.DataminerDeploy.csproj

Co-authored-by: Michiel Oda <97458010+MichielOda@users.noreply.github.com>

janstaelensskyline committed:
Update Skyline.DataMiner.CICD.Tools.DataminerDeploy.Lib/CICD.Tools.DataminerDeploy.Lib.csproj

Co-authored-by: Michiel Oda <97458010+MichielOda@users.noreply.github.com>

janstaelensskyline committed:
Update Skyline.DataMiner.CICD.Tools.DataminerDeploy.Lib/CatalogService/CatalogServiceFactory.cs

Co-authored-by: Michiel Oda <97458010+MichielOda@users.noreply.github.com>

janstaelensskyline committed:
Apply review remarks

janstaelensskyline committed:
Update Skyline.DataMiner.CICD.Tools.DataminerDeploy.Lib/CatalogService/HttpCatalogService.cs

Co-authored-by: Michiel Oda <97458010+MichielOda@users.noreply.github.com>

janstaelensskyline committed:
Licenses Updated

MichielOda commented on CICD.Tools.DataminerDeploy/Program.cs:
```diff
@@ -0,0 +1,167 @@
+	using Skyline.DataMiner.CICD.Tools.DataMinerDeploy.Lib;
+
+	/// <summary>
+	/// Deploys a package to DataMiner from the catalog or directly..
```
```suggestion
	/// Deploys a package to DataMiner from the catalog or directly.
```

janstaelensskyline committed:
adjusted arguments to match Microsoft Guidelines

MichielOda commented on CICD.Tools.DataminerDeploy/Program.cs:
```diff
@@ -0,0 +1,167 @@
+
+			var artifactId = new Option<string>(
+			name: "--artifact-id",
+			description: "The unique cloud artifact identifier as returned from performing a catalog-upload. e.g. dmscript/f764389f-5404-4c32-9ac9-b54366a3d5e0")
```
Is this a fictional guid or an actual artifact?

janstaelensskyline replied:
It's the SkylineCommunications_AutomationScriptTest_SDK_DataAcq testing automationscript. Gets uploaded for the Skyline Communications organization only. So it should be safe.

MichielOda commented on CICD.Tools.DataminerDeploy/Program.cs:
```diff
@@ -0,0 +1,167 @@
+				IsRequired = true
+			};
+
+			var fromArtifact = new Command("from-artifact", "Deploys a specific package from a local .dmapp to a DataMiner agent.")
```
Maybe comment out the artifact parts as currently this will be blocked.

janstaelensskyline replied:
I removed everything concerning local artifacts from this branch. Including the readme & library items.
First release will be only for catalog due to SLNETTypes.dll issues with .netstandard for direct deployment through this .NET6.0 tool.

MichielOda commented on CICD.Tools.DataminerDeploy/README.md:
````diff
@@ -0,0 +1,85 @@
+The most basic command will allow deployment of an artifact using the artifact identifier returned from performing an upload using ["dataminer-catalog-upload"](https://www.nuget.org/packages/Skyline.DataMiner.CICD.Tools.CatalogUpload).
+
+```console
+dataminer-package-deploy from-catalog --artifact-id "dmscript/f764389f-5404-4c32-9ac9-b54366a3d5e0" --dm-catalog-token "cloudConnectedToken"
````
Same comment as before, is this a fictional guid?

janstaelensskyline replied:
It's the SkylineCommunications_AutomationScriptTest_SDK_DataAcq testing automationscript. Gets uploaded for the Skyline Communications organization only. So it should be safe.

MichielOda commented on Skyline.DataMiner.CICD.Tools.DataminerDeploy.Lib/CatalogService/HttpCatalogService.cs:
```diff
@@ -0,0 +1,137 @@
+		}
+
+		/// <summary>
+		/// Starts the deployment process on a DataMiner Server. You'll need to call
```
Stopped midway with writing? Param, returns & exceptions aren't filled in either.

janstaelensskyline replied:
Indeed, probably noticed halfway that this was an internal class.
I completed the comments.

MichielOda commented on Skyline.DataMiner.CICD.Tools.DataminerDeploy.Lib/Models/DeployedPackage.cs:
```diff
@@ -0,0 +1,12 @@
+﻿namespace Skyline.DataMiner.CICD.Tools.DataMinerDeploy.Lib
+{
+	public class DeployedPackage
```
Add xml documentation

janstaelensskyline replied:
done

janstaelensskyline committed:
Update CICD.Tools.DataminerDeploy/Program.cs

Co-authored-by: Michiel Oda <97458010+MichielOda@users.noreply.github.com>

janstaelensskyline committed:
Applied review remarks. Removed all LocalArtifact code.

janstaelensskyline committed:
Added devopsmetrics

MichielOda commented on CICD.Tools.DataminerDeploy/Program.cs:
```diff
@@ -67,101 +68,91 @@ public static async Task<int> Main(string[] args)
-				artifact = DeploymentFactory.Cloud(artifactId, dmCatalogToken, logger);
+			finally
+			{
+				if (!string.IsNullOrWhiteSpace(devopsMetricsMessage))
```
```suggestion
				if (!String.IsNullOrWhiteSpace(devopsMetricsMessage))
```

janstaelensskyline committed:
Update CICD.Tools.DataminerDeploy/Program.cs

Co-authored-by: Michiel Oda <97458010+MichielOda@users.noreply.github.com>
